File: backend/internal/models/monthly_attendance_request.go
package models

import "time"

/*
 * 〇 月次勤怠申請
 *
 * 従業員が対象月の勤怠を月次申請し、
 * 管理者が承認・否認するための親テーブル。
 *
 * このテーブルに入れるもの：
 * 	・対象ユーザー
 * 	・対象年月
 * 	・月次申請状態
 * 	・申請日時
 * 	・申請メモ
 * 	・承認者
 * 	・承認日時
 * 	・否認理由
 * 	・取り下げ日時
 * 	・取り下げ理由
 *
 * このテーブルに入れないもの：
 * 	・日別の予定区分
 * 	・日別の予定時間
 * 	・日別の実績区分
 * 	・日別の実績時間
 * 	・日別交通費
 * 	・休憩時間
 *
 * 理由：
 * 	AttendanceDay は日別勤怠データだけを管理する。
 * 	MonthlyAttendanceRequest は、その月全体の申請・承認状態だけを管理する。
 *
 * 状態：
 * 	・PENDING  = 申請中
 * 	・APPROVED = 承認済み
 * 	・REJECTED = 否認済み
 * 	・CANCELED = 取り下げ済み
 *
 * 補足：
 * 	未申請はレコードなしで表現する。
 * 	そのため NOT_SUBMITTED はDBには保存しない。
 */
type MonthlyAttendanceRequest struct {
	ID uint `gorm:"primaryKey" json:"id"`

	// 対象ユーザーID
	UserID uint `gorm:"not null;index" json:"userId"`

	// 対象年
	TargetYear int `gorm:"not null;index" json:"targetYear"`

	// 対象月
	TargetMonth int `gorm:"not null;index" json:"targetMonth"`

	// 月次申請状態
	// 例：PENDING, APPROVED, REJECTED, CANCELED
	Status string `gorm:"type:varchar(30);not null" json:"status"`

	// 申請メモ
	RequestMemo *string `gorm:"type:text" json:"requestMemo"`

	// 申請日時
	RequestedAt *time.Time `json:"requestedAt"`

	// 承認者ID
	ApprovedBy *uint `json:"approvedBy"`

	// 承認日時
	ApprovedAt *time.Time `json:"approvedAt"`

	// 否認理由
	RejectedReason *string `gorm:"type:text" json:"rejectedReason"`

	// 否認日時
	RejectedAt *time.Time `json:"rejectedAt"`

	// 取り下げ理由
	CanceledReason *string `gorm:"type:text" json:"canceledReason"`

	// 取り下げ日時
	CanceledAt *time.Time `json:"canceledAt"`

	// 論理削除フラグ
	IsDeleted bool `gorm:"not null;default:false" json:"isDeleted"`

	// 作成日時
	CreatedAt time.Time `json:"createdAt"`

	// 更新日時
	UpdatedAt time.Time `json:"updatedAt"`

	// 論理削除日時
	DeletedAt *time.Time `json:"deletedAt"`
}

File: frontend/src/types/admin/monthlyAttendanceRequest.ts
/*
 * 管理者 月次勤怠申請 Type
 *
 * バックエンドの admin MonthlyAttendanceRequestResponse に対応する。
 *
 * 注意：
 * ・月次申請状態は日別勤怠や月次通勤定期ではなく、この型で管理する
 * ・勤怠日の保存とは別物
 * ・月次勤怠全体保存APIとは別物
 * ・管理者APIでは対象ユーザーIDを targetUserId で送る
 * ・管理者側では月次申請状態による勤怠編集ロックを行わない
 * ・承認者IDはフロントから送らず、バックエンドでJWTから取得する
 */

export type MonthlyAttendanceRequestStatus =
  | "NOT_SUBMITTED"
  | "PENDING"
  | "APPROVED"
  | "REJECTED"
  | "CANCELED";

export type MonthlyAttendanceRequest = {
  id: number | null;

  targetUserId: number;

  targetYear: number;
  targetMonth: number;

  status: MonthlyAttendanceRequestStatus;

  exists: boolean;

  editable: boolean;
  canSubmit: boolean;
  canCancel: boolean;
  canApprove: boolean;
  canReject: boolean;

  requestMemo: string | null;
  requestedAt: string | null;

  approvedBy: number | null;
  approvedAt: string | null;

  rejectedReason: string | null;
  rejectedAt: string | null;

  canceledReason: string | null;
  canceledAt: string | null;

  createdAt: string | null;
  updatedAt: string | null;
};

/*
 * 月次勤怠申請一覧検索 Row
 *
 * POST /admin/monthly-attendance-requests/search
 *
 * 注意：
 * ・申請一覧画面用
 * ・ユーザー情報と対象月の月次申請状態を1行にまとめた型
 * ・未申請の場合、monthlyAttendanceRequest.status は NOT_SUBMITTED
 * ・未申請の場合、monthlyAttendanceRequest.id は null
 */
export type MonthlyAttendanceRequestListRow = {
  targetUserId: number;

  userName: string;
  email: string;

  departmentId: number | null;
  departmentName: string | null;

  targetYear: number;
  targetMonth: number;

  monthlyAttendanceRequest: MonthlyAttendanceRequest;
};

/*
 * 月次勤怠申請一覧検索
 *
 * POST /admin/monthly-attendance-requests/search
 *
 * 注意：
 * ・対象年月、ユーザーのフリーワード、申請状態で絞り込む
 * ・statuses は複数選択可能
 * ・未申請も検索対象にする場合は NOT_SUBMITTED を含める
 * ・includeDeletedUsers は退職者/削除済みユーザーも対象に含める場合に true
 * ・offset / limit / hasMore でページングする
 */
export type SearchMonthlyAttendanceRequestsRequest = {
  keyword: string;

  targetYear: number;
  targetMonth: number;

  statuses: MonthlyAttendanceRequestStatus[];

  includeDeletedUsers: boolean;

  offset: number;
  limit: number;
};

export type SearchMonthlyAttendanceRequestsResponse = {
  monthlyAttendanceRequests: MonthlyAttendanceRequestListRow[];
  hasMore: boolean;
};

/*
 * 月次勤怠申請状態取得
 *
 * POST /admin/monthly-attendance-requests/status
 *
 * 注意：
 * ・バックエンド側のルート名は search ではなく status
 * ・user側の命名に合わせて SearchMonthlyAttendanceRequestRequest / Response とする
 */
export type SearchMonthlyAttendanceRequestRequest = {
  targetUserId: number;
  targetYear: number;
  targetMonth: number;
};

export type SearchMonthlyAttendanceRequestResponse = {
  monthlyAttendanceRequest: MonthlyAttendanceRequest;
};

/*
 * 月次勤怠申請
 *
 * POST /admin/monthly-attendance-requests/submit
 */
export type SubmitMonthlyAttendanceRequestRequest = {
  targetUserId: number;
  targetYear: number;
  targetMonth: number;
  requestMemo: string | null;
};

export type SubmitMonthlyAttendanceRequestResponse = {
  monthlyAttendanceRequest: MonthlyAttendanceRequest;
};

/*
 * 月次勤怠申請取り下げ
 *
 * POST /admin/monthly-attendance-requests/cancel
 *
 * 注意：
 * ・バックエンド側のルート名は withdraw ではなく cancel
 * ・user側の命名に合わせて WithdrawMonthlyAttendanceRequestRequest / Response とする
 */
export type WithdrawMonthlyAttendanceRequestRequest = {
  targetUserId: number;
  targetYear: number;
  targetMonth: number;
  canceledReason: string | null;
};

export type WithdrawMonthlyAttendanceRequestResponse = {
  monthlyAttendanceRequest: MonthlyAttendanceRequest;
};

/*
 * 月次勤怠申請承認
 *
 * POST /admin/monthly-attendance-requests/approve
 */
export type ApproveMonthlyAttendanceRequestRequest = {
  targetRequestId: number;
};

export type ApproveMonthlyAttendanceRequestResponse = {
  monthlyAttendanceRequest: MonthlyAttendanceRequest;
};

/*
 * 月次勤怠申請否認
 *
 * POST /admin/monthly-attendance-requests/reject
 */
export type RejectMonthlyAttendanceRequestRequest = {
  targetRequestId: number;
  rejectedReason: string;
};

export type RejectMonthlyAttendanceRequestResponse = {
  monthlyAttendanceRequest: MonthlyAttendanceRequest;
};
